_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+categories:
+  - Torch
+  - R
+  - Time Series
+bibliography: bibliography.bib
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+You're reading the final post in a four-part introduction to time-series forecasting with `torch`. It's been the story of a quest for multiple-step prediction, and by now, we've seen three different approaches: forecasting in a loop, incorporating an MLP, and sequence-to-sequence models. Here's a quick recap.
+
+-   As one should when one sets out for an adventurous journey, we [started](https://blogs.rstudio.com/ai/posts/2021-03-10-forecasting-time-series-with-torch_1/) with an in-depth study of the tools at our disposal: recurrent neural networks (RNNs). We trained a model to predict the very next observation in line, and then, thought of a clever hack: How about we use this for multi-step prediction, feeding back individual predictions in a loop? The result , it turned out, was quite acceptable.
+
+-   [Then](https://blogs.rstudio.com/ai/posts/2021-03-11-forecasting-time-series-with-torch_2/), the adventure really started. We built our first model "natively" for multi-step prediction, relieving the RNN a bit of its workload and involving a second player, a tiny-ish multi-layer perceptron (MLP). Now, it was the MLP's task to project RNN output to several time points in the future. Although results were pretty satisfactory, we didn't stop there.
+
+-   Instead, we [applied](https://blogs.rstudio.com/ai/posts/2021-03-16-forecasting-time-series-with-torch_3/) to numerical time series a technique commonly used in natural language processing (NLP): sequence-to-sequence (*seq2seq*) prediction. While forecast performance was not much different from the previous case, we found the technique to be more intuitively appealing, since it reflects the *causal* relationship between successive forecasts.
+
+Today, we don't replace *seq2seq* by something completely different. Instead, we enrich it, adding in a new component: the so-called *attention* module. Originally introduced around 2014[^1], attention mechanisms have gained enormous traction, so much so that a recent paper title starts out "Attention is Not All You Need"[^2].
+
+[^1]: For important papers see: @BahdanauCB14, @VinyalsKKPSH14, @XuBKCCSZB15.
+
+[^2]: @2021arXiv210303404D, thus replying to @2017arXiv170603762V, whose 2017 paper "went viral" for stating the opposite.
+
+The idea is the following. In the classic encoder-decoder setup, the decoder gets "primed" with an encoder summary just a single time: the time it starts its forecasting loop. From then on, it's on its own. With attention, however, it gets to see the complete sequence of encoder outputs again every time it forecasts a new value. And more than that: Every time, it gets to zoom in on *those* outputs that seem *relevant for the current prediction step*.
```

Comment:
Maybe this section should get a header, since it's where you begin explaining the key concept of the post (and it will make it easier for someone to skip to the relevant section).

Reply:
I'm torn here ... the _what's more_, that is exactly the idiomatic phrase that didn't come to mind, but ...

>The idea is the following.

Yes, I know, I know ... but in a way, that's just my style ... and relatedly,

> Maybe this section should get a header, since it's where you begin explaining the key concept of the post (and it will make it easier for someone to skip to the relevant section).

Reply:
> I trust readers to assess the relevance (to them) of individual paragraphs

I don't think this is how most readers act. The Nielsen Norman report, [How People Read Online: The Eyetracking Evidence](https://www.nngroup.com/reports/how-people-read-web-eyetracking-evidence/) is long, but the [article here](https://totheweb.com/blog/b2b-content-marketing/how-people-read-on-the-web-a-guide-for-b2b-content-marketing/) is a quick summary.

Reply:
thanks for the link(s)! I just skimmed (;-)))) the latter, and yes, there's useful advice (part of which I should definitely take into account more often).
But, I do think the content category plays a role, as well.

>They drop-off for a variety of reasons, but far and away the number repellant is a wall of text.

For sure that's true in general, but not necessarily as much here ("a wall", sure, that's bad. But let's say "some occasionally wordy writing" instead.)

For some people, what might be even more repelling is a big amount of "non-text" (math, for example). So when you see a text about ML that's more than math and code, that could be a good thing, too ;-)

Not saying any of those findings are irrelevant, just that one might see some of them in context.